### ISSUE TITLE
Remove whitespace from edited lines only

### ISSUE THREAD

**Issue by spyder-issues-migrator** [spyder-ide/spyder#757]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2011-09-14T09:58:27Z_

Spyder Version:  2.1.0beta1 

What steps will reproduce the problem?

 1.Tools -> Preferences -> Editor -> Advanced Settings
2.'Automatically remove trailing whitespaces when saving files' What is the expected output? What do you see instead? Spyder poorly handles trailing whitespace problem: people are forced to waste time looking for differences in diff output when there are no semantic changes except trailing stuff. This is especially evil in case of version control history where you often need to annotate lines to find causes of problems, and every whitespace (or linefeed) change is an extra step, which becomes very irritating at times.

Most version control users don't want editors to automatically remove _all_ trailing spaces from edited files. This touches parts of file that they didn't edit. Eclipse suxx at that. What they want is to remove trailing whitespaces only from the lines they've actually edited.

If this mode is implemented, the option mentioned above can be removed. If somebody needs to remove all trailings whitespaces from the whole file, it is very easy to do from Python console.

_Original issue: http://code.google.com/p/spyderlib/issues/detail?id=757_

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2011-09-14T07:59:02Z_

**Labels:** -Type-Defect Type-Enhancement

**Comment by spyder-issues-migrator** [comment]
_From [ccordoba12](https://code.google.com/u/ccordoba12/) on 2011-10-08T10:33:00Z_

**Labels:** Cat-Editor

**Comment by spyder-issues-migrator** [comment]
_From [pmav99@gmail.com](https://code.google.com/u/pmav99@gmail.com/) on 2011-11-02T09:21:15Z_

I think that the option should not be removed. It is still useful in some cases. Perhaps a new option would be more appropriate.

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2011-11-02T10:18:47Z_

@-pmav99: to me the option to remove _all_ trailing whitespace in file is mostly project specific, and removal from edited lines should just be the default

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2011-11-02T10:35:43Z_

I tend to agree with pmav99:  implementing the "removal from edited lines" feature should be independent from the existing option that we should keep.

**Status:** Accepted  
  **Labels:** Milestone-Release2.2

**Comment by spyder-issues-migrator** [comment]
_From [pmav99@gmail.com](https://code.google.com/u/pmav99@gmail.com/) on 2012-01-10T03:04:02Z_

Anyway, I think that a simple change in the VCS workflow, eliminates the need for this feature.

Just resaving the file without making any modifications, thus erasing all trailing whitespace, and commiting afterwards results in a "clean" file. Afterwards you can make your modifications, without viewing differences all over the place.

Just my 2 cents.

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2012-01-10T03:42:53Z_

Great development tools "help developers write great software while staying out of the way." (c) Trac

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2012-01-10T03:45:21Z_

I mean great tools help people to write software and not impose their own way of doing things. I use annotations very often, and I know what am I asking for. Everybody is free to add their own 0.02 cents, of course. =)

**Comment by spyder-issues-migrator** [comment]
_From [pmav99@gmail.com](https://code.google.com/u/pmav99@gmail.com/) on 2012-01-12T09:40:45Z_

I tend to consider trailing whitespace a bad practice. The reasoning behind this is exactly the problems with the diff files that you described.

In regard of existing projects, I believe that it is very easy to write a script that will recursively open all source files and remove trailing whitespace. Afterwards you just commit the clean files to your repo and you have solved the problem once and for all. So afterall, this is not really a change to the chosen workflow. It is just a single step in order to improve general source code quality.

I don't say that this kind of feature, cannot become handy, but since there is a simple and cleaner way to solve the problem, perhaps this should be the way to go. So adding an option to open recursively all the files of a folder (maybe allowing to specify certain extensions) and removing their trailing whitespace is probably easier to implement and eventually more useful.

With a bit of guidance, I believe I could provide a patch for this feature.

regards,
Panos

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2012-01-12T12:45:19Z_

There are projects where some parts of the code were not touched for ages,
won't be touched and better not be touched at all. Age of the file is a
very useful metrics when researching problems in such projects. You need to
get your hands dirty with support of some large big project to understand
that whitespace cleanups and bulk code formatting can be really really
annoying (really really). In such projects you really want to reformat or
remove space only from the lines you are thoroughly read, understood
and going to modify.

In other words this "simple and cleaner way" doesn't make the job of
software archaeologist any better - it makes it a lot worse.

**Comment by spyder-issues-migrator** [comment]
_From [pmav99@gmail.com](https://code.google.com/u/pmav99@gmail.com/) on 2012-01-12T15:39:45Z_

You are right of course. I don't really have experience with such projects. Or, to be more accurate, my experience with legacy code involves only fortran code that isn't under version control, so it doesn't really matter in this context.

But the existence of such old projects, doesn't justify choosing the removal of trailing whitespace of only the edited lines as the default behavior for spyder. This doesn't mean that it is not a useful feature. I think that adding both features would be the best for spyder.

with regards,
Panos

**Comment by spyder-issues-migrator** [comment]
_From [techtonik@gmail.com](https://code.google.com/u/techtonik@gmail.com/) on 2012-01-12T19:23:00Z_

“Perfection is achieved perfection not when there is nothing left to add, but when there is nothing left to take away” – Antoine de Saint-Exupery

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2012-03-18T13:48:10Z_

**Labels:** -Milestone-Release2.2 Milestone-v2.2

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2012-03-18T13:58:49Z_

**Labels:** -Milestone-v2.2 MS-v2.2

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2012-03-18T14:14:48Z_

**Labels:** -Type-Enhancement Type-Enh

**Comment by spyder-issues-migrator** [comment]
_From [pierre.raybaut](https://code.google.com/u/pierre.raybaut/) on 2013-01-27T01:48:04Z_

Issue opened in 2011, nothing new in 2012. Closing it in 2013...

**Status:** WontFix
